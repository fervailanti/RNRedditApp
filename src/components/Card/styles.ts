import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  card: {
    flexDirection: 'row',
    backgroundColor: 'white',
    padding: 12,
    marginTop: 16,
    marginHorizontal: 16,
    borderRadius: 8,
    shadowColor: 'black',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 3,
  },
  image: {
    height: 80,
    width: 80,
    borderRadius: 4,
  },
  container: {
    flex: 1,
    marginLeft: 12,
    justifyContent: 'space-between',
  },
  top: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginBottom: 2,
  },
  bottom: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    marginTop: 2,
  },
  indicatorLast: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  indicator: {
    flexDirection: 'row',
    alignItems: 'center',
    marginRight: 12,
  },
  indicatorLabel: {
    fontSize: 12,
    marginRight: 2,
    fontWeight: 'bold',
    color: '#999',
  },
  indicatorIcon: {
    marginRight: 2,
  },
  indicatorText: {
    fontWeight: 'normal',
    fontSize: 12,
    color: '#999',
  },
  title: {
    fontSize: 14,
  },
  flex: {
    flex: 1,
  },
})
