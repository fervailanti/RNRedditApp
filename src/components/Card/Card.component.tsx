import { Icon } from '@components'
import { isHttpUrl } from '@utils/helpers'
import React from 'react'
import {
  Image,
  StyleProp,
  Text,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native'
import styles from './styles'

type Indicator = {
  flex?: boolean
  label?: string
  icon?: string
  text: string | number
}

type CardProps = {
  onPress?: () => void
  imageUrl?: string
  title?: string
  topIndicators?: Indicator[]
  bottomIndicators?: Indicator[]
}

const Card: React.FC<CardProps> = ({
  onPress,
  imageUrl,
  title,
  topIndicators,
  bottomIndicators,
}) => {
  const renderIndicators = (
    indicators: Indicator[],
    style: StyleProp<ViewStyle>,
  ) => (
    <View style={style}>
      {indicators.map((indicator, index) => {
        const isLast = index + 1 === indicators.length
        const indicatorStyle = isLast ? styles.indicatorLast : styles.indicator
        return (
          <View
            key={index}
            style={[indicatorStyle, indicator.flex && styles.flex]}>
            {!!indicator.icon && (
              <Icon name={indicator.icon} style={styles.indicatorIcon} />
            )}
            {!!indicator.label && (
              <Text style={styles.indicatorLabel}>{indicator.label}</Text>
            )}
            <Text numberOfLines={1} style={styles.indicatorText}>
              {indicator.text}
            </Text>
          </View>
        )
      })}
    </View>
  )

  const showImage = !!imageUrl && isHttpUrl(imageUrl)

  return (
    <TouchableOpacity style={styles.card} onPress={onPress}>
      {showImage && (
        <Image
          resizeMode="cover"
          source={{ uri: imageUrl }}
          style={styles.image}
        />
      )}
      <View style={styles.container}>
        {topIndicators && renderIndicators(topIndicators, styles.top)}
        {title && (
          <Text numberOfLines={2} style={styles.title}>
            {title}
          </Text>
        )}
        {bottomIndicators && renderIndicators(bottomIndicators, styles.bottom)}
      </View>
    </TouchableOpacity>
  )
}

export default Card
