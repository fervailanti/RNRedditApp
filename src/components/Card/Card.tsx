import { Icon } from '@components'
import React from 'react'
import {
  Image,
  Pressable,
  StyleProp,
  Text,
  View,
  ViewStyle,
} from 'react-native'
import styles from './styles'

type Indicator = {
  flex?: boolean
  label?: string
  icon?: string
  text: string | number
}

type CardProps = {
  onPress?: () => void
  imageUrl?: string
  title?: string
  topIndicators?: Indicator[]
  bottomIndicators?: Indicator[]
}

const Card: React.FC<CardProps> = ({
  onPress,
  imageUrl,
  title,
  topIndicators,
  bottomIndicators,
}) => {
  const renderIndicators = (
    indicators: Indicator[],
    style: StyleProp<ViewStyle>,
  ) => (
    <View style={style}>
      {indicators.map((indicator, index) => {
        const isLast = index + 1 === indicators.length
        const indicatorStyle = isLast ? styles.indicatorLast : styles.indicator
        return (
          <View
            key={index}
            style={[indicatorStyle, indicator.flex && styles.flex]}>
            {!!indicator.icon && (
              <Icon name={indicator.icon} style={styles.indicatorIcon} />
            )}
            {!!indicator.label && (
              <Text style={styles.indicatorLabel}>{indicator.label}</Text>
            )}
            <Text style={styles.indicatorText}>{indicator.text}</Text>
          </View>
        )
      })}
    </View>
  )

  return (
    <Pressable style={styles.card} onPress={onPress}>
      {imageUrl && (
        <Image
          resizeMode="cover"
          source={{ uri: imageUrl }}
          style={styles.image}
        />
      )}
      <View style={styles.container}>
        {topIndicators && renderIndicators(topIndicators, styles.top)}
        {title && (
          <Text numberOfLines={2} style={styles.title}>
            {title}
          </Text>
        )}
        {bottomIndicators && renderIndicators(bottomIndicators, styles.bottom)}
      </View>
    </Pressable>
  )
}

export default Card
