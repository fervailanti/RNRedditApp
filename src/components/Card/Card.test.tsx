import { render } from '@testing-library/react-native'
import React from 'react'
import 'react-native'
import Card from './Card.component'

const props = {
  imageUrl: 'https://via.placeholder.com/100',
  title: 'Test Card',
  topIndicators: [{ label: 'T Label', text: 'T Text' }],
  bottomIndicators: [{ label: 'B Label', text: 'B Text' }],
  onPress: () => {},
}

describe('Components > Card', () => {
  it('<Card {...props} /> should render correctly', () => {
    const { toJSON } = render(<Card {...props} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('Given elements inside Card should be visible', () => {
    const { getByText } = render(<Card {...props} />)
    expect(getByText('Test Card')).toBeTruthy()
    expect(getByText('T Label')).toBeTruthy()
    expect(getByText('T Text')).toBeTruthy()
    expect(getByText('B Label')).toBeTruthy()
    expect(getByText('B Text')).toBeTruthy()
  })
})
