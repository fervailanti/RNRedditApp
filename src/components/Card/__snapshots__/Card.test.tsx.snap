// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > Card <Card {...props} /> should render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 8,
      "elevation": 3,
      "flexDirection": "row",
      "marginBottom": 16,
      "opacity": 1,
      "padding": 12,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.22,
      "shadowRadius": 2.22,
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "https://via.placeholder.com/100",
      }
    }
    style={
      Object {
        "borderRadius": 4,
        "height": 80,
        "marginRight": 12,
        "width": 80,
      }
    }
  />
  <View
    style={
      Object {
        "flex": 1,
        "height": 80,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "marginBottom": 2,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#999999",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginRight": 2,
            }
          }
        >
          T Label
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#999999",
              "fontSize": 12,
              "fontWeight": "normal",
            }
          }
        >
          T Text
        </Text>
      </View>
    </View>
    <Text
      numberOfLines={2}
      style={
        Object {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "500",
        }
      }
    >
      Test Card
    </Text>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "marginTop": 2,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            },
            undefined,
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "#999999",
              "fontSize": 12,
              "fontWeight": "bold",
              "marginRight": 2,
            }
          }
        >
          B Label
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#999999",
              "fontSize": 12,
              "fontWeight": "normal",
            }
          }
        >
          B Text
        </Text>
      </View>
    </View>
  </View>
</View>
`;
